internal/version/version.go

```diff
@@ -0,0 +1,110 @@
+/*
+Copyright (c) Edgeless Systems GmbH
+
+SPDX-License-Identifier: AGPL-3.0-only
+*/
+
+/*
+Package version provides functionality to parse and process semantic versions, as they are used in multiple components of Constellation.
+*/
+package version
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/edgelesssys/constellation/v2/internal/constants"
+	"golang.org/x/mod/semver"
+)
+
+// Version represents a semantic version.
+type Version struct {
+	Major int
+	Minor int
+	Patch int
+}
+
+// NewVersion returns a Version from a string.
+func NewVersion(version string) (Version, error) {
+	// ensure that semver has "v" prefix
+	if !strings.HasPrefix(version, "v") {
+		version = "v" + version
+	}
+
+	if !semver.IsValid(version) {
+		return Version{}, fmt.Errorf("invalid semver: %s", version)
+	}
+
+	version = semver.Canonical(version)
+
+	var major, minor, patch int
+	_, err := fmt.Sscanf(version, "v%d.%d.%d", &major, &minor, &patch)
+	if err != nil {
+		return Version{}, fmt.Errorf("parsing semver parts: %w", err)
+	}
+
+	return Version{
+		Major: major,
+		Minor: minor,
+		Patch: patch,
+	}, nil
+}
+
+// String returns the string representation of the version.
+func (v Version) String() string {
+	return fmt.Sprintf("v%d.%d.%d", v.Major, v.Minor, v.Patch)
+}
+
+// Compare compares two versions. It relies on the semver.Compare function internally.
+func (v Version) Compare(other Version) int {
+	return semver.Compare(v.String(), other.String())
+}
+
+// IsUpgradeTo returns if a version is an upgrade to another version.
+// It checks if the version of v is greater than the version of other and allows a drift of at most one minor version.
+func (v Version) IsUpgradeTo(other Version) bool {
+	return v.Compare(other) > 0 && v.Major == other.Major && v.Minor-other.Minor <= 1
+}
+
+// CompatibleWithBinary returns if a version is compatible version of the current built binary.
+// It checks if the version of the binary is equal or greater than the current version and allows a drift of at most one minor version.
+func (v Version) CompatibleWithBinary() bool {
+	binaryVersion, err := NewVersion(constants.VersionInfo)
+	if err != nil {
+		return false
+	}
+
+	return v.Equals(binaryVersion) || binaryVersion.IsUpgradeTo(v)
+}
+
+// NextMinor returns the next minor version in the format "vMAJOR.MINOR".
+func (v Version) NextMinor() string {
+	return fmt.Sprintf("v%d.%d", v.Major, v.Minor+1)
+}
+
+// Equals returns if two versions are equal.
+func (v Version) Equals(other Version) bool {
```

Comment:
Duplicate to `Compare`? Or is there another reason for implementing this that I am not seeing?